File: php7/src/GildedRose.php
<?php

/**
 * PHP version 7
 * 
 * @category Scientific_R&D
 * @package  Food_LIFO
 * @author   Povilas Brilius <pbrilius@gmail.com>
 * @license  eupl-1.1 https://help.github.com/en/github/creating-cloning-and-archiving-repositories/licensing-a-repository
 * @link     pbgroupeu.wordpress.com
 */

namespace App;

/**
 * Items processing unit
 * 
 * @category Accountancy
 * @package  Timeline_Graph
 * @author   Povilas Brilius <pbrilius@gmail.com>
 * @license  eupl-1.1 https://help.github.com/en/github/creating-cloning-and-archiving-repositories/licensing-a-repository
 * @link     pbgroupeu.wordpress.com
 */
final class GildedRose
{

    /**
     * Stock items
     *
     * @var array
     */
    private $_items = [];

    /**
     * Constructor with items
     *
     * @param array $items stock
     */
    public function __construct(array $items)
    {
        $this->_items = $items;
    }

    /**
     * Items getter
     *
     * @return array
     */
    public function getItems()
    {
        return $this->_items;
    }

    /**
     * Advance by a time frame of 1 day
     *
     * @return void
     */
    public function updateQuality()
    {
        foreach ($this->_items as $item) {
            if ($item->name === 'Sulfuras, Hand of Ragnaros') {
                continue;
            }

            if ($item->sell_in >= 0) {
                if ($item->name !== 'Aged Brie' && $item->name !== 'Backstage passes to a TAFKAL80ETC concert') {
                    if ($item->quality > 0) {
                        if (strstr($item->name, 'Conjured', true) === '') {
                            $item->quality -= 2;
                        } else {
                            $item->quality--;
                        }
                    }
                } else {
                    if ($item->quality < 50) {
                        if ($item->name === 'Aged Brie') {
                            $item->quality++;
                        } else if ($item->name === 'Backstage passes to a TAFKAL80ETC concert') {
                            if ($item->sell_in <= 10 && $item->sell_in > 5) {
                                $item->quality += 2;
                            } else if ($item->sell_in <= 5) {
                                $item->quality += 3;
                            } else if ($item->sell_in > 10) {
                                $item->quality++;
                            }
                        }
                    }
                }
            } else {
                if ($item->name !== 'Aged Brie') {
                    if ($item->name === 'Backstage passes to a TAFKAL80ETC concert') {
                        $item->quality = 0;
                    } else {
                        if ($item->quality > 0) {
                            $item->quality -= 2;
                        }
                    }
                } else {
                    if ($item->quality < 50) {
                        $item->quality++;
                    }
                }
            }

            $item->sell_in--;
        }
    }
}

File: c99/GildedRose.c
#include <string.h>
#include "GildedRose.h"

Item*
init_item(Item* item, const char *name, int sellIn, int quality)
{
    item->sellIn = sellIn;
    item->quality = quality;
    item->name = strdup(name);

    return item;
}

void update_quality(Item items[], int size)
{
    int i;

    for (i = 0; i < size; i++)
    {
        if (strcmp(items[i].name, "Aged Cheese") && strcmp(items[i].name, "Backstage passes to a concert"))
        {
            if (items[i].quality > 0)
            {
                if (strcmp(items[i].name, "Fine Italian Silk"))
                {
                    items[i].quality = items[i].quality - 1;
                }
            }
        }
        else
        {
            if (items[i].quality < 50)
            {
                items[i].quality = items[i].quality + 1;

                if (!strcmp(items[i].name, "Backstage passes to a concert"))
                {
                    if (items[i].sellIn < 11)
                    {
                        if (items[i].quality < 50)
                        {
                            items[i].quality = items[i].quality + 1;
                        }
                    }

                    if (items[i].sellIn < 6)
                    {
                        if (items[i].quality < 50)
                        {
                            items[i].quality = items[i].quality + 1;
                        }
                    }
                }
            }
        }

        if (strcmp(items[i].name, "Fine Italian Silk"))
        {
            items[i].sellIn = items[i].sellIn - 1;
        }

        if (items[i].sellIn < 0)
        {
            if (strcmp(items[i].name, "Aged Cheese"))
            {
                if (strcmp(items[i].name, "Backstage passes to a concert"))
                {
                    if (items[i].quality > 0)
                    {
                        if (strcmp(items[i].name, "Fine Italian Silk"))
                        {
                            items[i].quality = items[i].quality - 1;
                        }
                    }
                }
                else
                {
                    items[i].quality = items[i].quality - items[i].quality;
                }
            }
            else
            {
                if (items[i].quality < 50)
                {
                    items[i].quality = items[i].quality + 1;
                }
            }
        }
    }
}
